patches/server/0966-Expand-PlayerDeathEvent-API.patch

```diff
@@ -0,0 +1,62 @@
++        org.bukkit.event.entity.PlayerDeathEvent event = CraftEventFactory.callPlayerDeathEvent(this, loot, PaperAdventure.asAdventure(defaultMessage), defaultMessage.getString(), keepInventory, flag); // Paper - Adventure
+         // Paper start - cancellable death event
+         if (event.isCancelled()) {
+             // make compatible with plugins that might have already set the health in an event listener
+@@ -922,13 +922,14 @@ public class ServerPlayer extends Player {
+         }
+ 
+         net.kyori.adventure.text.Component deathMessage = event.deathMessage() != null ? event.deathMessage() : net.kyori.adventure.text.Component.empty(); // Paper - Adventure
++        Component deathScreenMessage = PaperAdventure.asVanilla(event.deathScreenMessage() != null && event.shouldShowDeathMessage() ? event.deathScreenMessage() : net.kyori.adventure.text.Component.empty()); // Paper - death screen api
+ 
+-        if (deathMessage != null && deathMessage != net.kyori.adventure.text.Component.empty() && flag) { // Paper - Adventure // TODO: allow plugins to override?
++        if (deathMessage != null && deathMessage != net.kyori.adventure.text.Component.empty() && event.shouldShowDeathMessage()) { // Paper - Adventure
+             Component ichatbasecomponent = PaperAdventure.asVanilla(deathMessage); // Paper - Adventure
+ 
+-            this.connection.send(new ClientboundPlayerCombatKillPacket(this.getCombatTracker(), ichatbasecomponent), PacketSendListener.exceptionallySend(() -> {
++            this.connection.send(new ClientboundPlayerCombatKillPacket(this.getCombatTracker(), deathScreenMessage), PacketSendListener.exceptionallySend(() -> { // Paper - death screen api
+                 boolean flag1 = true;
+-                String s = ichatbasecomponent.getString(256);
++                String s = deathScreenMessage.getString(256); // Paper - death screen api
+                 MutableComponent ichatmutablecomponent = Component.translatable("death.attack.message_too_long", Component.literal(s).withStyle(ChatFormatting.YELLOW));
+                 MutableComponent ichatmutablecomponent1 = Component.translatable("death.attack.even_more_magic", this.getDisplayName()).withStyle((chatmodifier) -> {
+                     return chatmodifier.withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, ichatmutablecomponent));
+@@ -948,7 +949,7 @@ public class ServerPlayer extends Player {
+                 this.server.getPlayerList().broadcastSystemMessage(ichatbasecomponent, false);
+             }
+         } else {
+-            this.connection.send(new ClientboundPlayerCombatKillPacket(this.getCombatTracker(), CommonComponents.EMPTY));
++            this.connection.send(new ClientboundPlayerCombatKillPacket(this.getCombatTracker(), deathScreenMessage)); // Paper - death screen api
```

Comment:
This diff shouldn't be needed. If the death message shouldn't be sent/shown, then an empty component should always be passed to the `ClientboundPlayerCombatKillPacket`.

Reply:
Well what about case, when you don't want to set the deathMessage to other players  but want to show the death message to the player dying? Thinking of it, I am afraid the diff cannot be as simple as that, because this would miss the check for the length of the message...

Reply:
You're billing the shouldShowDeathMessages method as a way to give per-death control over the gamerule, it should follow the rules of that gamerule. And that gamerules is either all or nothing. If you want to add further control over things, then what that method is should be rethought or repurposed.

Reply:
> because this would miss the check for the length of the message...
the length check is inside the if statement that's below it. It's not skipping it.

Reply:
I was more talking about the case, when you set the deathMessage null, deathScreenMessage not null and shouldShowDeathMessage to true. In this case it should only show the death message player to the player dying.

Reply:
Wouldn't this then be a change in behavior for existing plugins? Say I already have a listener that sets the death message to null. Now, the player who died suddenly does get a message on their screen when they didn't before.

Reply:
You're right, this would break the original behavior. My original implementation would not break it, but as lynxplay pointed out it would be counterintuitive and would make a case when you don't want to show the message to the dying player but to everyone else yes more complicated (you would have to send empty component instead of null).

Ok, another possible solution is to deprecate current death message methods (that would set message for both the one dying and everyone else) and introduce a new pair of functions to set/get the message for everyone and the one dying separately. I can't think of a cleaner solution... What do you think?

Reply:
Should I go with the abovementioned solution? Deprecating the old message and introducing a pair of new ones?